### ISSUE TITLE
Reconnecting a monitor sets the wrong ID

### ISSUE THREAD

**Issue by zjeffer** [hyprwm/Hyprland#2601]
I have 3 monitors, with IDs (0, 1, 2). If I disconnect the monitor with ID=1, and then reconnect it, the new ID will be 3, so my monitor IDs will now be (0, 2, 3). This doesn't make sense to me. Why not reuse ID 1?

I noticed the logic is in here: https://github.com/hyprwm/Hyprland/blob/41358c6fb579e440ccb72bb79a7bfb859944b3a6/src/Compositor.cpp#L1700-L1708

Why does it not use the minimum available ID?

**Comment by vaxerski** [comment]
cuz thats what I coded lol, relying on IDs is weird anyways, why not use port names / descriptions?

**Comment by zjeffer** [comment]
I forked @Duckonaut's [split-monitor-workspaces plugin](https://github.com/zjeffer/split-monitor-workspaces) so it would automatically assign a fixed number of persistent workspaces to every monitor, based on the ID:

Monitor with ID=0 gets workspaces 1-5
Monitor with ID=1 gets workspaces 6-10
Monitor with ID=2 gets workspaces 11-15

This way, I can have a bspwm-like experience where I can switch workspaces by putting my mouse on a monitor, and simply pressing `Super+<n>` to go to the `n`th workspace on that focused monitor.

The problem is that waybar needs to show the same 5 workspaces that the plugin assigned: so the bar on monitor with ID=1 should also show workspaces 6-10. I can simply use the same workspace name calculation as in the plugin: `workspace_name = ID * count + index`. I forked Waybar as well to make this possible: https://github.com/zjeffer/Waybar/tree/persistent_ws_improvements

It all works perfectly, except for the fact that when I disconnect and reconnect a monitor, my workspace names will now be:

* Monitor 0: 1-5
* Monitor 1: 16-20
* Monitor 2: 11-15

because the ID was not reused. It still works fine, but it's just ugly. Disconnecting and reconnecting again would from now on always assign ID=3 to monitor 1, it will never use ID=1. This just doesn't make sense to me. It would be better if the next available ID would be the minimum one. 

---

Aside from all this, I still don't get why I have to jump through so many hoops to get persistent workspaces on every monitor, with keybinds like how it works in bspwm. I feel like that guy from the Diablo mobile game ["Do you guys not have ~~phones~~ multi monitor setups?"](https://www.youtube.com/watch?v=50KBNQe5hTM)

**Comment by MightyPlaza** [comment]
You could do everything you did by using port instead of the ID

**Comment by MightyPlaza** [comment]
https://github.com/Alexays/Waybar/wiki/Module:-Workspaces#persistent-workspaces
Doesn't waybar already do that?
Also I still don't get the need to use the plugin
http://wiki.hyprland.org/Configuring/Monitors/#binding-workspaces-to-a-monitor

**Comment by zjeffer** [comment]
> https://github.com/Alexays/Waybar/wiki/Module:-Workspaces#persistent-workspaces
> Doesn't waybar already do that?

That option doesn't work very well, and that's the reason for my fork. I created a draft PR here with more info if you're interested: https://github.com/Alexays/Waybar/pull/2251. The main problem is that it doesn't work very dynamically: if you connect a monitor that is not defined in your waybar config, it won't add the workspaces I want. 

> Also I still don't get the need to use the plugin
> http://wiki.hyprland.org/Configuring/Monitors/#binding-workspaces-to-a-monitor 

The plugin is mainly for the keybinds. It automatically assigns workspaces to monitors, and sets up the keybinds so that Super+n always switches to the nth workspace on the focused monitor. I don't want to change my config every time I connect to a new monitor.

Basically, I want the following:
* Every monitor should have a fixed number of persistent workspaces
* My Waybar should show the monitor's workspaces. Clicking the workspace should switch to it properly.
* Using super+<n> should switch to the nth workspace on the focused monitor
* Disconnecting and connecting a monitor should work seamlessly and create new persistent workspace bindings
* I should not have to manually specify monitor names in my Waybar or Hyprland config. This should all happen automatically.

My forks of the plugin and Waybar fix this problem perfectly. The only issue I have is an aesthetic one: the list of monitor IDs should in my opinion not have any gaps in them.

**Comment by zjeffer** [comment]
Did not mean to close

**Comment by MightyPlaza** [comment]
"Every monitor should have a fixed number of persistent workspaces" -> can be done by binding a workspace to a monitor

"My Waybar should show the monitor's workspaces." -> Default waybar shows monitor non empty workspaces, you can add all the "fixed" workspaces with the persistent workspaces option

"Using super+<n> should switch to the nth workspace on the focused monitor" -> not yet possible, but doable with a simple script

"Disconnecting and connecting a monitor should work seamlessly and create new persistent workspace bindings" -> On monitor disconnect your workspaces would move to one of the other monitors and you wouldn't be able to access them due to the request above. If you want to move windows to another monitor and access them after monitor disconnect you would have to make a script that would define the behaviour you want. What windows should be move? Should a window that was moved from a monitor then closed and opened again still go to that monitor? Again all this is not necessary if you do not want to acess the windows. (forgot to mention that on monitor replug the workspaces move back to the original monitor, at least for me)

"I should not have to manually specify monitor names in my Waybar or Hyprland config. This should all happen automatically." -> you have to have some criterion for identifying monitors, the only one that's currently supported on hyprland is the port (DP-1), other specs like monitor desc don't work for all config options.

And that a long text enjoy reading

**Comment by zjeffer** [comment]
I typed up a whole response to your comment but I think it got deleted or not posted properly :( I think it maybe got autodeleted because I tagged Vaxry

Here's the short version of what I wrote:

1. binding a workspace to a monitor does not happen automatically. The plugin takes care of this
2. Again, not automatic. The waybar fork takes care of this
3. The script in this case is the plugin
4. This is something I haven't thought about yet, and it is indeed an issue. Right now, if I disconnect a monitor, the workspace + its windows are moved to another monitor, and this is how Waybar looks:

![example](https://github.com/hyprwm/Hyprland/assets/4633209/6d2f76e1-f5a9-495a-a9f9-589c41253ab7)

The workspace is clickable but of course not accessible with super+n keybinds. A better solution would be to move the windows to a (preferably empty) workspace on one of the remaining monitors. 

How does native hyprland take care of monitors with workspaces being disconnected/reconnected?

5. Sorry, here I meant workspacenames, not monitor names. I of course set the monitors with `monitor=<name>, etc` in my hyprland config.

---

Either way, this is kind of off topic to this issue. But thank you for your feedback.

I can implement the change myself, I'd just like to know if Vaxry agrees with this or not, before I start coding. If not, I can close this issue as intended behaviour.

**Comment by MightyPlaza** [comment]
> binding a workspace to a monitor does not happen automatically. The plugin takes care of this

You need to make a workspace rule

> Again, not automatic. The waybar fork takes care of this

I think the solution is to be able to blacklist workpaces on the waybar (oposite of persistent workspaces but similar design)
Maybe PR it

> How does native hyprland take care of monitors with workspaces being disconnected/reconnected? 

Moves the workspaces to an active monitor, and moves them back on monitor reconnect (not sure about the details)

> I should not have to manually specify ~monitor~ workspace names in my Waybar or Hyprland config. This should all happen automatically.

Again I don't see the problem here, just 10 rules for each monitor you'd have to write once, but yes maybe there should be a way to apply worskpace rules to workspaces from a list all at once instead of 1 rule for each.

**Comment by zjeffer** [comment]
I should also mention that I use a Dell docking station at work, and I'm kind of preparing my setup so I can use Hyprland with it. That's one of the reasons I'm putting so much effort into getting this as automatic as possible. The problem is that the docking station creates about 8 different outputs (2 of which have monitors connected), and for some reason they sometimes change the output names (I think after firmware upgrades). 

For instance, the docking station (with 2 monitors connected) creates the outputs [DP-1-1, DP-1-2, ... , DP-1-8] (I can see this on X11 with xrandr). On one day I might connect my laptop to it, and the monitors are assigned to the DP-1-1 and DP-1-2 outputs, and another day they might be assigned to the DP-1-3 and DP-1-4 outputs. I don't know why this happens. 

With the native implementation of Hyprland and Waybar, I would have to write dozens of rules in my config. If I connect to another persons docking station (that might give different outputs), I don't have to screw around with both my Hyprland and Waybar config files just to set up my workspaces. 

With my forks, I just specify the amount of workspaces any monitor should have (you can specify a default amount with the * character instead of monitor name), and that's it.

**Comment by MightyPlaza** [comment]
That's a issue a lot of people have been having (monitor ports changing).
Most people use a script to handle the monitor names -> port and the making of the widow/workspace rules, but again this is the hacky solution.
An actual solution is adding support to monitor desc to other workspace and window rules on hyprland itself, again good PR idea and probably just copy paste code from monitor conf parsing to workspace/window rule parsing.

The waybar part would be about the same, adding support to monitor desc

**Comment by zjeffer** [comment]
What do you mean by monitor desc? The EDID? The dimensions?

**Comment by MightyPlaza** [comment]
in hyprctl monitors  the description param

**Comment by MightyPlaza** [comment]
https://github.com/hyprwm/Hyprland/pull/2670
Allows you to use monitor description to bind the workspaces to a monitor without using the port
